libs/pubsub/query/query.go

```diff
@@ -1,527 +1,327 @@
-		filteredValue := numRegex.FindString(value)
-
-		// try our best to convert value from tags to float64
-		v, err := strconv.ParseFloat(filteredValue, 64)
-		if err != nil {
-			return false, fmt.Errorf("failed to convert value %v from event attribute to float64: %w", filteredValue, err)
+// matchesEvent reports whether c matches the given event.
+func (c condition) matchesEvent(event types.Event) bool {
+	vs, tagEqualsType := c.findAttr(event)
+	if len(vs) == 0 {
+		// As a special case, a condition tag that exactly matches the event type
+		// is matched against an empty string. This allows existence checks to
+		// work for type-only queries.
+		if tagEqualsType {
+			return c.match("")
 		}
+		return false
+	}
 
-		switch op {
-		case OpLessEqual:
-			return v <= operandFloat64, nil
-		case OpGreaterEqual:
-			return v >= operandFloat64, nil
-		case OpLess:
-			return v < operandFloat64, nil
-		case OpGreater:
-			return v > operandFloat64, nil
-		case OpEqual:
-			return v == operandFloat64, nil
+	// At this point, we have candidate values.
+	for _, v := range vs {
+		if c.match(v) {
+			return true
 		}
+	}
+	return false
+}
 
-	case reflect.Int64:
-		var v int64
-
-		operandInt := operand.Interface().(int64)
-		filteredValue := numRegex.FindString(value)
-
-		// if value looks like float, we try to parse it as float
-		if strings.ContainsAny(filteredValue, ".") {
-			v1, err := strconv.ParseFloat(filteredValue, 64)
-			if err != nil {
-				return false, fmt.Errorf("failed to convert value %v from event attribute to float64: %w", filteredValue, err)
-			}
-
-			v = int64(v1)
-		} else {
-			var err error
-			// try our best to convert value from tags to int64
-			v, err = strconv.ParseInt(filteredValue, 10, 64)
-			if err != nil {
-				return false, fmt.Errorf("failed to convert value %v from event attribute to int64: %w", filteredValue, err)
-			}
-		}
+func compileCondition(cond syntax.Condition) (condition, error) {
+	out := condition{tag: cond.Tag}
 
-		switch op {
-		case OpLessEqual:
-			return v <= operandInt, nil
-		case OpGreaterEqual:
-			return v >= operandInt, nil
-		case OpLess:
-			return v < operandInt, nil
-		case OpGreater:
-			return v > operandInt, nil
-		case OpEqual:
-			return v == operandInt, nil
-		}
+	// Handle existence checks separately to simplify the logic below for
+	// comparisons that take arguments.
+	if cond.Op == syntax.TExists {
+		out.match = func(string) bool { return true }
+		return out, nil
+	}
 
-	case reflect.String:
-		switch op {
-		case OpEqual:
-			return value == operand.String(), nil
-		case OpContains:
-			return strings.Contains(value, operand.String()), nil
-		}
+	// All the other operators require an argument.
+	if cond.Arg == nil {
+		return condition{}, fmt.Errorf("missing argument for %v", cond.Op)
+	}
 
+	// Precompile the argument value matcher.
+	argType := cond.Arg.Type
+	var argValue interface{}
+
+	switch argType {
+	case syntax.TString:
```

Comment:
Is the `T` prefix in the variable name supposed to stand for type? Is this a common convention?

Reply:
"T"oken, actually. I don't know if I can say it's "common", but having a prefix for enumerators is not-uncommon. I didn't think it was worth making a whole `token` package just to namespace these (adding another import at the usage site), but I could do that if you think it's better.